Description: Eliminates dependencies from static Qt
Author: Nicholas Guriev <guriev-ns@ya.ru>
Bug-Debian: https://bugs.debian.org/767418
Bug-Ubuntu: https://launchpad.net/bugs/1499516
Last-Update: 2016-06-18

--- telegram-desktop-0.9.51.orig/Telegram/MetaLang.pro
+++ telegram-desktop-0.9.51/Telegram/MetaLang.pro
@@ -5,11 +5,13 @@ CONFIG(debug, debug|release) {
     OBJECTS_DIR = ./../DebugIntermediateLang
     MOC_DIR = ./GeneratedFiles/Debug
     DESTDIR = ./../DebugLang
+    OUTPUT = ../DebugIntermediate/GeneratedFiles
 }
 CONFIG(release, debug|release) {
     OBJECTS_DIR = ./../ReleaseIntermediateLang
     MOC_DIR = ./GeneratedFiles/Release
     DESTDIR = ./../ReleaseLang
+    OUTPUT = ../ReleaseIntermediate/GeneratedFiles
 }
 
 CONFIG += plugin static c++11
@@ -27,4 +29,8 @@ HEADERS += \
     ./SourceFiles/_other/mlmain.h \
     ./SourceFiles/_other/genlang.h \
 
-include(qt_static.pri)
+INCLUDEPATH += "/usr/include/$$(DEB_HOST_MULTIARCH)/qt5/QtGui/$$[QT_VERSION]/QtGui" \
+               "/usr/include/$$(DEB_HOST_MULTIARCH)/qt5/QtCore/$$[QT_VERSION]/QtCore" \
+               "/usr/include/$$(DEB_HOST_MULTIARCH)/qt5"
+
+QMAKE_POST_LINK = $$DESTDIR/$$TARGET -lang_in ../../Telegram/Resources/langs/lang.strings -lang_out $$OUTPUT/lang_auto
--- telegram-desktop-0.9.51.orig/Telegram/SourceFiles/main.cpp
+++ telegram-desktop-0.9.51/Telegram/SourceFiles/main.cpp
@@ -25,10 +25,11 @@ Copyright (c) 2014-2016 John Preston, ht
 #include "localstorage.h"
 
 int main(int argc, char *argv[]) {
-#ifndef Q_OS_MAC // Retina display support is working fine, others are not.
-	QCoreApplication::setAttribute(Qt::AA_DisableHighDpiScaling, true);
-#endif // Q_OS_MAC
-
+#ifdef Q_OS_LINUX64
+	Application::addLibraryPath("/usr/lib/x86_64-linux-gnu/qt5/plugins");
+#elif defined Q_OS_LINUX32
+	Application::addLibraryPath("/usr/lib/i386-linux-gnu/qt5/plugins");
+#endif
 	settingsParseArgs(argc, argv);
 	if (cLaunchMode() == LaunchModeFixPrevious) {
 		return psFixPrevious();
--- telegram-desktop-0.9.51.orig/Telegram/SourceFiles/profile/profile_widget.cpp
+++ telegram-desktop-0.9.51/Telegram/SourceFiles/profile/profile_widget.cpp
@@ -128,7 +128,7 @@ void Widget::onScroll() {
 	auto windowHandle = window()->windowHandle();
 	auto globalPoint = QCursor::pos();
 	auto localPoint = windowHandle->mapFromGlobal(globalPoint);
-	QMouseEvent ev(QEvent::MouseMove, localPoint, localPoint, globalPoint, Qt::NoButton, QGuiApplication::mouseButtons(), QGuiApplication::keyboardModifiers(), Qt::MouseEventSynthesizedByApplication);
+	QMouseEvent ev(QEvent::MouseMove, localPoint, localPoint, globalPoint, Qt::NoButton, QGuiApplication::mouseButtons(), QGuiApplication::keyboardModifiers());
 	ev.setTimestamp(getms());
 
 	QGuiApplication::sendEvent(windowHandle, &ev);
--- telegram-desktop-0.9.51.orig/Telegram/SourceFiles/stdafx.cpp
+++ /dev/null
@@ -1,45 +0,0 @@
-/*
-This file is part of Telegram Desktop,
-the official desktop version of Telegram messaging app, see https://telegram.org
-
-Telegram Desktop is free software: you can redistribute it and/or modify
-it under the terms of the GNU General Public License as published by
-the Free Software Foundation, either version 3 of the License, or
-(at your option) any later version.
-
-It is distributed in the hope that it will be useful,
-but WITHOUT ANY WARRANTY; without even the implied warranty of
-MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
-GNU General Public License for more details.
-
-In addition, as a special exception, the copyright holders give permission
-to link the code of portions of this program with the OpenSSL library.
-
-Full license: https://github.com/telegramdesktop/tdesktop/blob/master/LICENSE
-Copyright (c) 2014-2016 John Preston, https://desktop.telegram.org
-*/
-#include "stdafx.h"
-#include <QtCore/QtPlugin>
-
-#ifdef Q_OS_WINRT
-//Q_IMPORT_PLUGIN(QWinRTIntegrationPlugin)
-//Q_IMPORT_PLUGIN(QWbmpPlugin)
-#elif defined Q_OS_WIN // Q_OS_WINRT
-Q_IMPORT_PLUGIN(QWindowsIntegrationPlugin)
-Q_IMPORT_PLUGIN(QWebpPlugin)
-#elif defined Q_OS_MAC // Q_OS_WIN
-Q_IMPORT_PLUGIN(QGenericEnginePlugin)
-Q_IMPORT_PLUGIN(QCocoaIntegrationPlugin)
-Q_IMPORT_PLUGIN(QDDSPlugin)
-Q_IMPORT_PLUGIN(QICNSPlugin)
-Q_IMPORT_PLUGIN(QICOPlugin)
-Q_IMPORT_PLUGIN(QTgaPlugin)
-Q_IMPORT_PLUGIN(QTiffPlugin)
-Q_IMPORT_PLUGIN(QWbmpPlugin)
-Q_IMPORT_PLUGIN(QWebpPlugin)
-#elif defined Q_OS_LINUX // Q_OS_LINUX
-Q_IMPORT_PLUGIN(QComposePlatformInputContextPlugin)
-Q_IMPORT_PLUGIN(QIbusPlatformInputContextPlugin)
-Q_IMPORT_PLUGIN(QFcitxPlatformInputContextPlugin)
-Q_IMPORT_PLUGIN(QWebpPlugin)
-#endif
--- telegram-desktop-0.9.51.orig/Telegram/SourceFiles/stdafx.h
+++ telegram-desktop-0.9.51/Telegram/SourceFiles/stdafx.h
@@ -22,10 +22,10 @@ Copyright (c) 2014-2016 John Preston, ht
 #define NOMINMAX // no min() and max() macro declarations
 #define __HUGE
 
-// Fix Google Breakpad build for Mac App Store version
-#ifdef Q_OS_MAC
+// Fix Google Breakpad build for Mac App Store version and for Linux Ubuntu
+#if defined Q_OS_MAC || defined Q_OS_LINUX
 #define __STDC_FORMAT_MACROS
-#endif // Q_OS_MAC
+#endif // Q_OS_MAC || Q_OS_LINUX
 
 #ifdef __cplusplus
 
--- telegram-desktop-0.9.51.orig/Telegram/SourceFiles/ui/text/text.cpp
+++ telegram-desktop-0.9.51/Telegram/SourceFiles/ui/text/text.cpp
@@ -30,6 +30,43 @@ Copyright (c) 2014-2016 John Preston, ht
 #include "boxes/confirmbox.h"
 #include "mainwindow.h"
 
+// This is a copy from Qt source file qtextengine.cpp for TextPainter::drawLine.
+// Temporary fix it here.
+// TODO: find a dynamic library which contains this methods.
+QTextItemInt::QTextItemInt(const QGlyphLayout &g, QFont *font, const QChar *chars_, int numChars, QFontEngine *fe, const QTextCharFormat &format)
+    : flags(0), justified(false), underlineStyle(QTextCharFormat::NoUnderline), charFormat(format),
+      num_chars(numChars), chars(chars_), logClusters(0), f(font),  glyphs(g), fontEngine(fe)
+{
+}
+
+// Fix up flags and underlineStyle with given info
+void QTextItemInt::initWithScriptItem(const QScriptItem &si)
+{
+    // explicitly initialize flags so that initFontAttributes can be called
+    // multiple times on the same TextItem
+    flags = 0;
+    if (si.analysis.bidiLevel %2)
+        flags |= QTextItem::RightToLeft;
+    ascent = si.ascent;
+    descent = si.descent;
+
+    if (charFormat.hasProperty(QTextFormat::TextUnderlineStyle)) {
+        underlineStyle = charFormat.underlineStyle();
+    } else if (charFormat.boolProperty(QTextFormat::FontUnderline)
+               || f->d->underline) {
+        underlineStyle = QTextCharFormat::SingleUnderline;
+    }
+
+    // compat
+    if (underlineStyle == QTextCharFormat::SingleUnderline)
+        flags |= QTextItem::Underline;
+
+    if (f->d->overline || charFormat.fontOverline())
+        flags |= QTextItem::Overline;
+    if (f->d->strikeOut || charFormat.fontStrikeOut())
+        flags |= QTextItem::StrikeOut;
+}
+
 namespace {
 
 const style::textStyle *_textStyle = nullptr;
--- telegram-desktop-0.9.51.orig/Telegram/SourceFiles/ui/text/text_block.cpp
+++ telegram-desktop-0.9.51/Telegram/SourceFiles/ui/text/text_block.cpp
@@ -333,7 +333,7 @@ TextBlock::TextBlock(const style::font &
 		QStackTextEngine engine(part, blockFont->f);
 		engine.itemize();
 
-		QTextLayout layout(&engine);
+		QTextLayout layout(part, blockFont->f);
 		layout.beginLayout();
 		layout.createLine();
 
--- telegram-desktop-0.9.51.orig/Telegram/Telegram.pro
+++ telegram-desktop-0.9.51/Telegram/Telegram.pro
@@ -1,6 +1,9 @@
 QT += core gui network widgets
 
-CONFIG += plugin static c++11
+CONFIG += plugin c++14
+
+DEFINES += TDESKTOP_DISABLE_AUTOUPDATE
+DEFINES += TDESKTOP_DISABLE_REGISTER_CUSTOM_SCHEME
 
 CONFIG(debug, debug|release) {
     DEFINES += _DEBUG
@@ -10,7 +13,6 @@ CONFIG(debug, debug|release) {
     DESTDIR = ./../Debug
 }
 CONFIG(release, debug|release) {
-    DEFINES += CUSTOM_API_ID
     OBJECTS_DIR = ./../ReleaseIntermediate
     MOC_DIR = ./GeneratedFiles/Release
     RCC_DIR = ./GeneratedFiles
@@ -29,28 +31,6 @@ linux {
     HEADERS += ./SourceFiles/pspecific_linux.h
 }
 
-codegen_style.target = style_target
-codegen_style.depends = FORCE
-codegen_style.commands = ./../codegen/Debug/codegen_style "-I./../../Telegram/Resources" "-I./../../Telegram/SourceFiles" "-o./GeneratedFiles/styles" all_files.style --rebuild
-
-codegen_numbers.target = numbers_target
-codegen_numbers.depends = ./../../Telegram/Resources/numbers.txt
-codegen_numbers.commands = ./../codegen/Debug/codegen_numbers "-o./GeneratedFiles" "./../../Telegram/Resources/numbers.txt"
-
-CONFIG(debug, debug|release) {
-codegen_numbers.commands = cd ../../Telegram && ./../Linux/codegen/Debug/codegen_numbers "-o./../Linux/DebugIntermediate/GeneratedFiles" "./Resources/numbers.txt" && cd ../Linux/DebugIntermediate
-}
-CONFIG(release, debug|release) {
-}
-
-codegen_lang.target = lang_target
-codegen_lang.depends = ./../../Telegram/Resources/langs/lang.strings
-codegen_lang.commands = mkdir -p ./GeneratedFiles && ./../DebugLang/MetaLang -lang_in ./../../Telegram/Resources/langs/lang.strings -lang_out ./GeneratedFiles/lang_auto
-
-QMAKE_EXTRA_TARGETS += codegen_style codegen_numbers codegen_lang
-
-PRE_TARGETDEPS += style_target numbers_target lang_target
-
 unix {
     linux-g++:QMAKE_TARGET.arch = $$QMAKE_HOST.arch
     linux-g++-32:QMAKE_TARGET.arch = x86
@@ -73,7 +53,6 @@ SOURCES += \
     ./GeneratedFiles/styles/style_overview.cpp \
     ./GeneratedFiles/styles/style_profile.cpp \
     ./SourceFiles/main.cpp \
-    ./SourceFiles/stdafx.cpp \
     ./SourceFiles/apiwrap.cpp \
     ./SourceFiles/app.cpp \
     ./SourceFiles/application.cpp \
@@ -276,7 +255,7 @@ HEADERS += \
     ./SourceFiles/core/click_handler_types.h \
     ./SourceFiles/core/observer.h \
     ./SourceFiles/core/vector_of_moveable.h \
-	./SourceFiles/core/version.h \
+    ./SourceFiles/core/version.h \
     ./SourceFiles/data/data_abstract_structure.h \
     ./SourceFiles/data/data_drafts.h \
     ./SourceFiles/dialogs/dialogs_common.h \
@@ -385,10 +364,6 @@ HEADERS += \
   ./SourceFiles/pspecific_mac.h
 }
 
-SOURCES += \
-  ./ThirdParty/minizip/zip.c \
-  ./ThirdParty/minizip/ioapi.c
-
 CONFIG += precompile_header
 
 PRECOMPILED_HEADER = ./SourceFiles/stdafx.h
@@ -400,7 +375,7 @@ CONFIG(release, debug|release) {
     QMAKE_CXXFLAGS_RELEASE -= -O2
     QMAKE_CXXFLAGS_RELEASE += -Ofast -flto -fno-strict-aliasing -g
     QMAKE_LFLAGS_RELEASE -= -O1
-    QMAKE_LFLAGS_RELEASE += -Ofast -flto -g -rdynamic -static-libstdc++
+    QMAKE_LFLAGS_RELEASE += -Ofast -flto -g -rdynamic
 }
 # Linux 32bit fails Release link with Link-Time Optimization: virtual memory exhausted
 unix {
@@ -412,41 +387,38 @@ unix {
     }
 }
 CONFIG(debug, debug|release) {
-	QMAKE_LFLAGS_DEBUG += -g -rdynamic -static-libstdc++
+    QMAKE_LFLAGS_DEBUG += -g -rdynamic
 }
 
-include(qt_static.pri)
+INCLUDEPATH += "/usr/include/$$(DEB_HOST_MULTIARCH)/qt5/QtGui/$$[QT_VERSION]/QtGui" \
+               "/usr/include/$$(DEB_HOST_MULTIARCH)/qt5/QtCore/$$[QT_VERSION]/QtCore" \
+               "/usr/include/$$(DEB_HOST_MULTIARCH)/qt5"
 
-INCLUDEPATH += \
-               /usr/local/include\
-               /usr/local/include/opus\
-               ./SourceFiles\
-               ./GeneratedFiles\
-               ./ThirdParty/minizip\
-               ./../../Libraries/breakpad/src
+INCLUDEPATH += "./SourceFiles" \
+               "./GeneratedFiles"
 
+INCLUDEPATH += "/usr/include"
+INCLUDEPATH += "/usr/include/opus"
+INCLUDEPATH += "/usr/include/breakpad"
 INCLUDEPATH += "/usr/include/libappindicator-0.1"
+INCLUDEPATH += "/usr/include/libappindicator3-0.1"
 INCLUDEPATH += "/usr/include/gtk-2.0"
 INCLUDEPATH += "/usr/include/glib-2.0"
-INCLUDEPATH += "/usr/lib/x86_64-linux-gnu/glib-2.0/include"
-INCLUDEPATH += "/usr/lib/i386-linux-gnu/glib-2.0/include"
+INCLUDEPATH += "/usr/lib/$$(DEB_HOST_MULTIARCH)/glib-2.0/include"
 INCLUDEPATH += "/usr/include/cairo"
 INCLUDEPATH += "/usr/include/pango-1.0"
-INCLUDEPATH += "/usr/lib/x86_64-linux-gnu/gtk-2.0/include"
-INCLUDEPATH += "/usr/lib/i386-linux-gnu/gtk-2.0/include"
+INCLUDEPATH += "/usr/lib/$$(DEB_HOST_MULTIARCH)/gtk-2.0/include"
 INCLUDEPATH += "/usr/include/gdk-pixbuf-2.0"
 INCLUDEPATH += "/usr/include/atk-1.0"
-
 INCLUDEPATH += "/usr/include/dee-1.0"
 INCLUDEPATH += "/usr/include/libdbusmenu-glib-0.4"
+INCLUDEPATH += "/usr/include/minizip"
 
 LIBS += -ldl -llzma -lopenal -lavformat -lavcodec -lswresample -lswscale -lavutil -lopus -lva
-LIBS += $${QT_TDESKTOP_PATH}/plugins/platforminputcontexts/libcomposeplatforminputcontextplugin.a \
-        $${QT_TDESKTOP_PATH}/plugins/platforminputcontexts/libibusplatforminputcontextplugin.a \
-        $${QT_TDESKTOP_PATH}/plugins/platforminputcontexts/libfcitxplatforminputcontextplugin.a
-LIBS += /usr/local/lib/libz.a
-LIBS += /usr/local/lib/libxkbcommon.a
-LIBS += ./../../../Libraries/breakpad/src/client/linux/libbreakpad_client.a
+LIBS += -lz
+LIBS += -lminizip
+LIBS += -lbreakpad_client
+LIBS += -lcrypto -lssl
 
 RESOURCES += \
     ./Resources/telegram.qrc \
--- telegram-desktop-0.9.51.orig/Telegram/build/qmake/codegen_numbers/codegen_numbers.pro
+++ telegram-desktop-0.9.51/Telegram/build/qmake/codegen_numbers/codegen_numbers.pro
@@ -7,10 +7,12 @@ CONFIG -= app_bundle
 CONFIG(debug, debug|release) {
     OBJECTS_DIR = ./
     DESTDIR = ./../../../codegen/Debug
+    OUTPUT = ../../../DebugIntermediate/GeneratedFiles
 }
 CONFIG(release, debug|release) {
     OBJECTS_DIR = ./
     DESTDIR = ./../../../codegen/Release
+    OUTPUT = ../../../ReleaseIntermediate/GeneratedFiles
 }
 
 INCLUDEPATH += ./../../../SourceFiles
@@ -43,3 +45,5 @@ HEADERS += \
 ./../../../SourceFiles/codegen/numbers/options.h \
 ./../../../SourceFiles/codegen/numbers/parsed_file.h \
 ./../../../SourceFiles/codegen/numbers/processor.h
+
+QMAKE_POST_LINK = $$DESTDIR/$$TARGET -o $$OUTPUT ../../../../Telegram/Resources/numbers.txt
--- telegram-desktop-0.9.51.orig/Telegram/build/qmake/codegen_style/codegen_style.pro
+++ telegram-desktop-0.9.51/Telegram/build/qmake/codegen_style/codegen_style.pro
@@ -7,10 +7,12 @@ CONFIG -= app_bundle
 CONFIG(debug, debug|release) {
     OBJECTS_DIR = ./
     DESTDIR = ./../../../codegen/Debug
+    OUTPUT = ../../../DebugIntermediate/GeneratedFiles
 }
 CONFIG(release, debug|release) {
     OBJECTS_DIR = ./
     DESTDIR = ./../../../codegen/Release
+    OUTPUT = ../../../ReleaseIntermediate/GeneratedFiles
 }
 
 INCLUDEPATH += ./../../../SourceFiles
@@ -49,3 +51,5 @@ HEADERS += \
 ./../../../SourceFiles/codegen/style/processor.h \
 ./../../../SourceFiles/codegen/style/sprite_generator.h \
 ./../../../SourceFiles/codegen/style/structure_types.h
+
+QMAKE_POST_LINK = $$DESTDIR/$$TARGET -I ../../../../Telegram/Resources -I ../../../../Telegram/SourceFiles -o $$OUTPUT/styles ../../../../Telegram/Resources/all_files.style --rebuild
